Remove every other element

mojtabafayyaz opened the pull request:
---
name: solution review 
about: A template PR for code review with a checklist
---

<!--
  make this PR easy to find:

  - assign: yourself
  - link it to the corresponding challenge issue on the project board
-->

<!-- describe the challenge you will solve -->

# Behavior

## Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations

## Strategy

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function's strategy _is not_ described in the documentation
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

MadiMalik commented on .vscode/extensions.json:
why did you make this change?

MadiMalik commented on .vscode/settings.json:
Can you explain these changes?

MadiMalik reviewed (changes requested):
Hey Mojtaba,
Your code is good, but there’s room for improvement. You could add more test cases to check the function’s behavior with an empty list as a boundary case and how it handles duplicate values. Also please explain the changes you made in .vscode/extensions.json file

mojtabafayyaz committed:
Restore remove_every_other.py and test_remove_every_other.py

mojtabafayyaz committed:
Formatted test_remove_every_other.py with Black

mojtabafayyaz committed:
Added test cases for empty list and list with duplicates

mojtabafayyaz committed:
Updated test case for handling duplicates in remove_every_other function

mojtabafayyaz commented:
> Hey Mojtaba, Your code is good, but there’s room for improvement. You could add more test cases to check the function’s behavior with an empty list as a boundary case and how it handles duplicate values. Also please explain the changes you made in .vscode/extensions.json file

### Hi Madiha,

Thank you so much for your valuable feedback!

### Test Case Updates:
1. **Boundary Case for Empty List:**
   - I've added a test case to handle the empty list scenario, ensuring that the function behaves as expected even when no elements are present.

2. **Handling Duplicate Values:**
   - I’ve also added a test case to check how the function behaves with duplicate values. This helps confirm that the function correctly removes every second element, including duplicates, without any unexpected behavior.

### Regarding .vscode/extensions.json:
- The changes to the .vscode/extensions.json file were included mistakenly in my earlier commit. I’ve since cleaned them up and made sure they’re removed from the pull request before submission. I apologize for the oversight and confusion.

Thank you again for your helpful comments, and please feel free to let me know if there’s anything else I can improve.

Best regards,  
Mojtaba

MadiMalik reviewed (changes requested):
please add a module docstring to the test file as well

MadiMalik commented:
Thank you for cleaning up the .vscode/extensions.json file and adding additional test cases to your test file. I have one more suggestion: could you please add a module docstring to the test file as well? It’s one of the requirements mentioned in the list.

mojtabafayyaz committed:
Add module docstring to test_remove_every_other.py

mojtabafayyaz commented:
> please add a module docstring to the test file as well

Dear Madiha,  I just added a module docstring to the test file.

MadiMalik reviewed (approved):
Thanks for the docstring, Mojtaba

MadiMalik committed:
Merge branch 'main' into remove_every_other

MadiMalik reviewed (approved):
thanks for adding docstring
